sentinel/src/schedule/outbound.js

```diff
@@ -0,0 +1,232 @@
+      .catch(err => {
+        if (err.statusCode === 404) {
+          logger.error(`CouchDB config key 'medic-credentials/${key}' has not been populated. See the Outbound documentation.`);
+        }
+
+        // Throw it regardless so the process gets halted, we just error above for higher specificity
+        throw err;
+      });
+
+// Returns a list of tuples of [queueDoc, actualDoc]
+const queued = () =>
+  db.sentinel.allDocs({
+    startkey: 'task:outbound:',
+    endkey: 'task:outbound:\ufff0',
+    include_docs: true
+  })
+  .then(results => {
+    const queues = results.rows.map(r => r.doc);
+
+    return db.medic.allDocs({
+      keys: queues.map(q => q.doc_id),
+      include_docs: true
+    }).then(results => {
+      const docs = results.rows.map(r => r.doc);
+      return lineage.hydrateDocs(docs);
+    }).then(docs => queues.map((q, idx) => [q, docs[idx]]));
+  });
+
+// Maps a source document to a destination format using the given push config
+const map = (doc, conf) => {
+  const srcParams = srcData => {
+    if (typeof srcData === 'string') {
+      return {
+        path: srcData,
+        required: true
+      };
+    } else {
+      return {
+        path: srcData.path,
+        expr: srcData.expr,
+        required: !srcData.optional
+      };
+    }
+  };
+
+  const toReturn = {};
+
+  Object.keys(conf.mapping).forEach(dest => {
+    const {
+      path, expr, required
+    } = srcParams(conf.mapping[dest]);
+
+    let srcValue;
+
+    if (expr) {
+      try {
+        srcValue = vm.runInNewContext(expr, {doc: doc});
+      } catch (err) {
+        throw Error(`Mapping error for ${conf.key}/${dest} JS error on source document: ${doc._id}: ${err}`);
+      }
+    } else {
+      srcValue = objectPath.get({doc: doc}, path);
+    }
+
+    if (required && srcValue === undefined) {
+      throw Error(`Mapping error for ${conf.key}/${dest}: cannot find ${path} on source document`);
+    }
+
+    if (srcValue) {
+      objectPath.set(toReturn, dest, srcValue);
+    }
+  });
+
+  return toReturn;
+};
+
+// Attempts to send a given payload using a given push config
+const send = (payload, conf) => {
+  const sendOptions = {
+    method: 'POST',
+    url: urlJoin(conf.destination.base_url, conf.destination.path),
+    body: payload,
+    json: true
+  };
+
+  const auth = () => {
+    const authConf = conf.destination.auth;
+
+    if (!authConf) {
+      return Promise.resolve();
+    }
+
+    if (!authConf.type) {
+      return Promise.reject(Error('No auth.type, either declare the type or omit the auth property'));
+    }
+
+    if (authConf.type.toLowerCase() === 'basic') {
+      return credential(authConf['password_key'])
+        .then(password => {
+          sendOptions.auth = {
+            username: authConf.username,
+            password: password,
+            sendImmediately: true
+          };
+        });
+    }
+
+    if (authConf.type.toLowerCase() === 'muso-sih') {
```

Comment:
I definitely get where you're coming from, and the less partner specific code we have lying around or mixed in the better!

However, thinking about the pros and cons that I can see, I do still favour leaving it there for now, because:
 - We don't really have a framework for or a history of having partner-specific code, so to separate we'd have to probably forge a reasonably complicated path
 - Having two separate modules and one encapsulate another isn't free in terms of readability and complexity, comparatively to just having one
 - Currently it's honestly not _that_ mixed in, it's a pretty small if block (and associated documentation) that is still eminently deletable (though no doubt that could become less true over time)
 - I think we're still very much on the fence as to how long this code will live for anyway, and the use that it's going to get. I can very much imagine it going live, having a couple of very minor tweaks (better logging etc) and then being dropped for an external tool that is friendlier

I think if we end up having to bake any more weirdness in than just a different auth type we should definitely look at what you're talking about though.

Reply:
Sounds good